pkg/oauth/oauth.go

```diff
@@ -73,24 +80,16 @@ func (s *server) Authorize(authorizePage echo.HandlerFunc) echo.HandlerFunc {
 		ar.UserData = userData{UserIdentifiers: session.UserIdentifiers}
 		client := ttnpb.Client(ar.Client.(osinClient))
 		if !clientHasGrant(&client, ttnpb.GRANT_AUTHORIZATION_CODE) {
-			resp.SetError(osin.E_INVALID_CLIENT, "OAuth client does not have authorization code grant")
-			oauth2.FinishAuthorizeRequest(resp, req, ar)
-			return s.output(c, resp)
+			return errClientMissingGrant.WithAttributes("grant", "authorization_code")
```

Comment:
How is `E_INVALID_CLIENT` propagated to the caller?

Reply:
Previously these errors just resulted in a 500 error (the issue). Now we return an explicit error in the same way as all other errors. The error is rendered into the HTML page and then picked up by the webui code.

Reply:
OK I see, but with this explicit call to `SetError()` with a code I was under the impression it's some application layer transport of error codes, i.e. as part of OAuth and not just HTTP error codes

Reply:
Not for these internal errors. Some errors are indeed handled by the OAuth protocol (with an error param in the redirect), but if the OAuth client is invalid, we shouldn't redirect anyway.

Reply:
Ok, it turns out that the OAuth spec in fact allows redirecting on these errors, I'll revert the deletes.